### PR TITLE
Change ConnectionUpdate to PATCH behavior

### PR DESCRIPTION
**Pull request by pmossman**
## What
Addresses https://github.com/airbytehq/airbyte-cloud/issues/2064

## How
Changes behavior of `/connections/update` and `/web_backend/connections/update` routes to only modify connection properties that are present on the incoming request object. Any property on the incoming request object that is `null` indicates a connection property that should remain unchanged.

This will allow API callers to make small changes to connections without needing to pass in the entire connection object (which is particularly painful in that the entire catalog would need to be passed in for every single update).

This reduces the risk of using the connection update API, because if a field is accidentally left blank, it will no longer wipe out whatever was stored. 

This PR also assumes that callers will no longer use the old `schedule` property, and instead use the new `scheduleType` and `scheduleData` properties. This is important because with the old `schedule` property, `null` meant that the connection should be made 'manual'. This would be problematic with the patch-style implementation, because the server would have no way of distinguishing between a request that wants to set the schedule to `manual`, versus a request that wants to leave the schedule unchanged. Luckily, the front-end just switched over to the new schedule fields, so this PR adds some extra validation that the old `schedule` field is no longer being used for updates.

~Finally, this PR supports incremental catalog patches, which means the front-end only needs to include updated streams in the catalog field of the request. Any streams that aren't included in the request will remain unchanged in the connection's catalog. To remove a stream from the catalog, the front-end just needs to set `selected` to `false` in that stream's config.~
Update: I discovered that there are some cases where we rely on the full-replace behavior of updating connection catalogs, namely, when a schema is refreshed and a stream has been dropped, the front-end sends a catalog update that doesn't include the dropped streams. This is incompatible with stream-level patching, so this PR has been updated to treat the catalog like any other field - if present, replace the connection's catalog with the new value. if not present, leave the connection's catalog alone.

## Remaining / TODO
I've modified existing tests to cover the new behavior, but I will likely add some additional tests that cover the edges, especially around catalog patches and schedules.

I'll want to sync with front-end directly to test this out and make sure it lines up with expectations.

@timroes , @krishnaglick I'm not expecting you to review the backend logic, mostly just wanted to tag for awareness and coordination of testing. I _think_ these changes should be fully backwards compatible with your existing logic, since you send the full connection object for updates today. But it might be a good idea to start testing out what it would look like for the front-end to send partial patch updates, to make sure it works how we all expect.

## Recommended reading order
1. ConnectionsHandler.java
2. WebBackendConnectionsHandler.java (which calls ConnectionsHandler.updateConnection)
3. The rest, mostly just small supporting changes

### CODE REVIEW

File: airbyte-api/src/main/openapi/config.yaml
Comment (lmossman) [inline airbyte-api/src/main/openapi/config.yaml:3377]: I'm a little confused why we are using `operations` in this `webBackendConnectionUpdate` endpoint, but we are using `operationIds` in the `connectionUpdate` endpoint. Is there a reason we don't want these to be consistent?

I'm also a bit confused as to why we have two "connection update" endpoints at all. Is the non-webBackend one just used by API users or something? What is the value we get out of having two different endpoints, instead of just consolidating them into one?

---
EDIT: This comment lower down in the code helped me understand this a bit better:
```
   * As a convenience to the front-end, this endpoint also creates new operations present in the
   * request, and bundles those newly-created operationIds into the connection update.
```

But it still seems like it would also be easier for API users to not have to create/update operations elsewhere and instead just do it in the updateConnection endpoint like the webBackend one does?

**Review by lmossman — COMMENTED** [review]
I just had a couple questions about the API endpoints themselves, but otherwise all of the implementation changes you made look right to me! Thanks for all of the comments you left, they are really helpful for reviewing.

I have not tested these changes out myself yet

File: airbyte-api/src/main/openapi/config.yaml
Comment (pmossman) [reply airbyte-api/src/main/openapi/config.yaml:3377]: I think the main idea here is that the public-facing endpoints like `connectionUpdate` should remain as "pure" as possible, meaning we only modify one resource at a time. I think we *want* API users to first use the `operationCreate` endpoint, and then pass the returned IDs into the `connectionUpdate` endpoint. This way, users who are creating systems and pipelines powered by our API are using very clear, discrete building blocks.

The `web_backend` routes violate this idea for the sake of performance and convenience

File: airbyte-api/src/main/openapi/config.yaml
Comment (davinchia) [inline airbyte-api/src/main/openapi/config.yaml:1987]: +1 should definitely add this before merging in.

File: airbyte-api/src/main/openapi/config.yaml
Comment (davinchia) [reply airbyte-api/src/main/openapi/config.yaml:3377]: Yeap, the general approach we have on the apis are:
1) pure resource routes - for programmatic access.
2) client specific routes (there is a better industry buzzword for this) - essentially convenience routes for various frontends. Today we only have the `webbackend` prefix.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/helpers/ConnectionScheduleHelper.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/helpers/ConnectionScheduleHelper.java:55]: for my understanding, why populate this?

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java:379]:
```suggestion
    final List<UUID> updatedOperationIds = updateOperations(connectionRead, webBackendConnectionPatch);
```

Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java:379]: to make it clearer

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java:394]: thinking out loud: the section on resetting is long, what about pulling it out into a private method named `resetStreamsIfNeeded` or something along those lines.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java:530]: Love the comments!

File: airbyte-server/src/main/java/io/airbyte/server/handlers/helpers/ConnectionScheduleHelper.java
Comment (pmossman) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/helpers/ConnectionScheduleHelper.java:55]: I talked about this with @mfsiega-airbyte a bit, right now we still have some places in code that read the legacy column. Our core temporal workflow code doesn't, but some extraneous places do, like [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-scheduler/scheduler-persistence/src/main/java/io/airbyte/scheduler/persistence/job_tracker/TrackingMetadata.java#L41) for example.

I didn't want to expand the scope of this PR too much by also trying to eliminate every place in our code that reads from the column. We're still relying on "dual-write" for schedules, so this line makes sure that we still "dual-write" to the legacy column even for patch requests that don't specify the `schedule` property.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:350]: throw a RTE?

**Review by davinchia — COMMENTED** [review]
Did a quick pass focusing on the logic. Looks good so far @pmossman .

What is the test situation like? Want to make sure the patch functionality is well-tested and saw only 1 or 2 additional tests. Want to confirm that before approving.

File: airbyte-api/src/main/openapi/config.yaml
Comment (timroes) [reply airbyte-api/src/main/openapi/config.yaml:3377]: I think one of the reasons for web backend specific endpoints is also, that we often add "additional behavior" into them, e.g. figuring out if we should reset streams after updating, and in general if it would make sense to merge (in GET APIs) more resources together, so that the FE can have them in one call instead of two, we'd often change them (thus also no backward compatibility on them).

I do think there might be a reason for this "non pure" GET APIs, so we can optimize loading performance by having APIs that return us exactly what the FE needs even if it comes from separate APIs. Regarding the PATCH/POST (update) APIs, I feel maybe we could actually try to get them merged back into the "regular APIs" and have potential "additional behavior" just available via additional flags, e.g. as we already have with `skipReset` for the connection update, so we could tell the regular POST API which checks we'll need over the pure "saving of the resource", which most likely should be the default behavior.

**Comment by pmossman** [comment]
@davinchia I added some additional logic to preserve a sensible ordering to the catalog, because our merge logic uses set operations that don't preserve the order. Would be good to get some eyes on the function I wrote for this to make sure it's followable.

Also added a bunch of unit tests around updates. Will look into the WebBackendConnectionHandler tests soon to see if we need any updates there, and will also address the smaller comments/suggestions you left soon

File: airbyte-api/src/main/openapi/config.yaml
Comment (pmossman) [reply airbyte-api/src/main/openapi/config.yaml]: I think we likely will at some point, for now I wanted to keep the change non-breaking and backwards compatible with front-end, but once this lands we could discuss moving to `patch` instead of `post`. Right now our server makes some assumptions about everything being `POST`, and it'd be a breaking change on the front-end that we'd need to coordinate closely, but worth doing imo at some point in the near future.

 Some conversation about this in an issue that describes the broad convention for our API: https://github.com/airbytehq/airbyte/issues/13698

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java
Comment (pmossman) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:336]: @timroes @krishnaglick I wonder if this will pose any problems for the front-end? My validation here is assuming that requests that include `scheduleData` will always include `scheduleType` as well. I think this would be a good thing to enforce to remove any ambiguity in the update, but there might be existing front-end code that will fail the validation

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java
Comment (pmossman) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:409]: @davinchia want to call attention to this logic, which is new since your last review. Basically, now that callers don't need to supply the entire catalog for every request, I want to make sure we preserve the ordering of the streams when patching the catalog. This helper is pretty verbose but I think it gets the job done

**Comment by pmossman** [comment]
@davinchia @krishnaglick friendly bump on this PR!

File: airbyte-api/src/main/openapi/config.yaml
Comment (lmossman) [reply airbyte-api/src/main/openapi/config.yaml:3283]: Got it, the motivation around having these as two separate endpoints makes sense. Though, I think there would be value in consolidating these through the use of flags as Tim suggested, so that we don't have duplication of API definitions which will inevitably lead to them becoming out of sync (e.g. an engineer makes a change to the base endpoint but forgets to change the webBackend one, causing a bug().

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:354]:
```suggestion
         // shouldn't be possible to reach this case
        default ->  throw new RuntimeException("Unrecognized scheduleType!");
```

Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:354]: I think it's possible to get rid of the brackets here

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java
Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:409]: Thanks!

Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:409]: Remind me, why do we care about preserving the order?

Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:409]:
```suggestion
  private static List<AirbyteStreamAndConfiguration> sortStreamsForPatchedCatalog(final List<AirbyteStreamAndConfiguration> unSortedMergedStreams,
```

Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:409]: otherwise merged suggests some sort of sorting

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:437]:
```suggestion
      // likewise, if stream2 is in the original but stream1 isn't, stream2 should come first
```

File: airbyte-api/src/main/openapi/config.yaml
Comment (davinchia) [reply airbyte-api/src/main/openapi/config.yaml:3283]: The counterpoint to Tim's proposed approach is the consolidated endpoints become complicated as they start serving too many user types. This isn't likely to be great for the Cloud Api user segment. Because of this, prefer to keep separate for now.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/helpers/ConnectionScheduleHelper.java
Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/helpers/ConnectionScheduleHelper.java:55]: ok!

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java:472]: can we call this `createAndUpdateOperations` to make clear this is also creating?

File: airbyte-server/src/test/java/io/airbyte/server/handlers/ConnectionsHandlerTest.java
Comment (davinchia) [inline airbyte-server/src/test/java/io/airbyte/server/handlers/ConnectionsHandlerTest.java:330]: it's probably worth throwing all the update connection patch related test into an inner test class so it's clearer these share a group category. See the `StreamConfigurationDiff` class in this class.

File: airbyte-server/src/test/java/io/airbyte/server/handlers/ConnectionsHandlerTest.java
Comment (davinchia) [inline airbyte-server/src/test/java/io/airbyte/server/handlers/ConnectionsHandlerTest.java:457]: curious, do the tests with the multiple streams test for stream sorting behaviour?

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java
Comment (davinchia) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:416]: just a thought: if we pull this out into it's own function, it'll be easier to test the sorting logic. Not sure how relevant this is given the existing connection handler tests.

**Review by davinchia — APPROVED** [review]
Looks good to me.

I left some suggestions and comments. None are blocking since this change set seems to be well tested.

I really appreciate the tests and the comments. Well done!

**Comment by krishnaglick** [comment]
Looking through this I was expecting a new api endpoint that takes a patch. It seems like the endpoint just handles partial data now? Regardless of verb?

**Comment by pmossman** [comment]
@krishnaglick yeah, we aren't following exact REST semantics here, we decided that the existing API endpoint is too dangerous to leave as-is, since neglecting to include a field will often wipe that field out in our persistence layer. So the goal is to update the existing endpoint such that it is backwards compatible with any existing calls, but treats missing/null properties as "leave them alone" rather than "clear them out".

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java
Comment (pmossman) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:409]: I don't think it's _strictly_ necessary, but I think it shows some user empathy to preserve the order when making changes (much easier to verify the change as a human looking at an update response, no unexpected order changes in UIs that show the catalog, etc) and it's also easier to write tests since we can use equality matchers on entire `ConnectionRead` objects, rather than needing to implement a custom matcher that allows the catalog to be unsorted, but wants equality for every other property.

File: airbyte-server/src/test/java/io/airbyte/server/handlers/ConnectionsHandlerTest.java
Comment (pmossman) [reply airbyte-server/src/test/java/io/airbyte/server/handlers/ConnectionsHandlerTest.java:457]: I think they do implicitly (they were failing until I added the sorting behavior) but I like your suggestion of writing a test explicitly for the sort function, I'll add that in before merging

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java
Comment (pmossman) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java:416]: Wrote a test for this whole method rather than just the sort, I think it better demonstrates the intent of the behavior while also covering the sort behavior itself

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java
Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java]: Makes sense. Can we leave a comment here explaining that? Want to make it clear this is a UX decision and not an application constraint.

File: airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java
Comment (davinchia) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/ConnectionsHandler.java]: Sounds good!

**Comment by krishnaglick** [comment]
> @krishnaglick yeah, we aren't following exact REST semantics here, we decided that the existing API endpoint is too dangerous to leave as-is, since neglecting to include a field will often wipe that field out in our persistence layer. So the goal is to update the existing endpoint such that it is backwards compatible with any existing calls, but treats missing/null properties as "leave them alone" rather than "clear them out".

Makes sense. I think the FE can make some performance improvements with how we handle editing connections with all this!

**Comment by pmossman** [comment]
@krishnaglick yep, as soon as we finalize this PR and get it merged, I'll want to work with your team to get https://github.com/airbytehq/airbyte-cloud/issues/2499 taken care of next.

Once we remove the catalog from the response of our `web_backend/connections/list` route, the performance of the connections page should improve. I imagine removing the catalog from that list response will be something we need to coordinate, right? If we just removed it on the backend, I'm sure some front-end code/components would break?

**Comment by pmossman** [comment]
By the way, the only thing I'm waiting for to merge this is a deploy to a dev environment so that I can do some manual verification of updates, especially around catalogs and schedules. I've had a bunch of interrupts and issues with our dev cluster deploys today but should be able to complete the verification tomorrow and get this finally merged, fingers crossed!

**Comment by pmossman** [comment]
This PR is currently deployed to https://dev-3-cloud.airbyte.io/ if anyone wants to do some manual verification that things still work! I've clicked around and it looks good so far

**Pull request merged**
